### PR TITLE
Unread notification performance.

### PR DESCRIPTION
**Pull request by erikjohnston**
This re-enables the unread notification functionality.

Major changes:

- Add `_filter_events_for_clients` to do filtering of events for multiple users more efficient.
- Remove the serialization of events as events now look sufficiently dict like.
- `BulkPushRuleEvaluator` does less things in the inner loops.
- `BulkPushRuleEvaluator` caches results of conditions on each event so that conditions that default conditions only get evaluated once per user in the room.
- Add `PushRuleEvaluatorForEvent` that caches some information for each event to make condition matching quicker.
- Only use regex matching if pattern has special glob characters in it.
- Change `synapse.push.base_rules` to have the base rules as constant dicts.

### CODE REVIEW

**Comment by erikjohnston** [comment]
matrixbot: test this please

File: synapse/storage/push_rule.py [hunk 1]
```diff
@@ -60,6 +60,27 @@ def get_push_rules_enabled_for_user(self, user_id):
...
+            )
+            txn.execute(sql, (user_id,))
+            return self.cursor_to_dict(txn)
+
+        return self.runInteraction(
+            "_get_push_rules_enabled_for_user", f
+        )
+
+    # @cachedList(cache=_get_push_rules_enabled_for_user.cache, list_name="user_ids",
+    #             num_args=1, inlineCallbacks=True)
```
Comment (NegativeMjark) [inline synapse/storage/push_rule.py]: Why is this commented out here?

File: synapse/push/push_rule_evaluator.py [hunk 2]
```diff
@@ -139,94 +172,141 @@ def actions_for_event(self, ev):
...
+
+        return False
+
+    def _get_value(self, dotted_key):
+        return self._value_cache.get(dotted_key, None)
+
+
+def _glob_to_matcher(glob):
+    """Takes a glob and returns a `func(string) -> bool`, which returns if the
+    string matches the glob. Assumes given string is lower case.
```
Comment (NegativeMjark) [inline synapse/push/push_rule_evaluator.py]: "Assumes that the string passed to the function is lowercase"?

File: synapse/push/push_rule_evaluator.py [hunk 3]
```diff
@@ -139,94 +172,141 @@ def actions_for_event(self, ev):
...
+
+    The matcher returned is either a simple string comparison for globs without
+    wildcards, or a regex matcher for globs with wildcards.
+    """
```
Comment (NegativeMjark) [inline synapse/push/push_rule_evaluator.py]: Is the ``glob`` a ``str``?

File: synapse/push/push_rule_evaluator.py [hunk 4]
```diff
@@ -139,94 +172,141 @@ def actions_for_event(self, ev):
...
+                x.group(1) and '^' or '',
+                x.group(2).replace(r'\\\-', '-')
+            )
+        ),
+        r,
+    )
+
+    r = r + "$"
+    r = re.compile(r)
+    return lambda value: r.match(value)
```
Comment (NegativeMjark) [inline synapse/push/push_rule_evaluator.py]: Why not just return ``r.match``?

File: synapse/storage/push_rule.py (hunk 1 above)
Comment (dbkr) [inline synapse/storage/push_rule.py]: Is this used, if the annotation here is commented out?

File: synapse/push/push_rule_evaluator.py [hunk 5]
```diff
@@ -139,94 +172,141 @@ def actions_for_event(self, ev):
...
-            # XXX: optimisation: cache our pattern regexps
-            if condition['key'] == 'content.body':
-                r = r'\b%s\b' % PushRuleEvaluator._glob_to_regexp(condition['pattern'])
-            else:
-                r = r'^%s$' % PushRuleEvaluator._glob_to_regexp(condition['pattern'])
-            val = _value_for_dotted_key(condition['key'], ev)
-            if val is None:
-                return False
-            return re.search(r, val, flags=re.IGNORECASE) is not None
+    def create(event, room_member_count):
```
Comment (NegativeMjark) [inline synapse/push/push_rule_evaluator.py]: Why is this a thing?
Why not do all this in the constructor?

**Comment by erikjohnston** [comment]
PTAL

**Comment by dbkr** [comment]
OK, I've mostly only looked at your 'Make notifications go quicker' but this lgtm.

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**